pkg/trace/api/api.go

```diff
@@ -589,7 +603,15 @@ func decodeRequest(req *http.Request, dest msgp.Decodable) error {
 	default:
 		// do our best
 		if err1 := json.NewDecoder(req.Body).Decode(dest); err1 != nil {
-			if err2 := msgp.Decode(req.Body, dest); err2 != nil {
+			buffer := bufferPool.Get().(*bytes.Buffer)
+			buffer.Reset()
```

Comment:
These two can go into a `newBuffer` func

Reply:
I see the benefit of factoring the code, but at the same time I like the fact that the `bufferPool.Get()` is explicit and visible in the same function than  `bufferPool.Put()` I think it makes the flow clearer and makes it less likely to copy the code and forget to Put back the buffer in the pool. What do you think?

Reply:
The general practice is a global variable for a `sync.Pool` and a `new*` constructor to get items from it and reset. That's what I've always been using and that's the form used in the Go standard library too (you can grep for both `sync.Pool` and `Pool.Get` to see for yourself if you'd like).

On that note, "best practice" can also be subjective, but I think in this case there is enough evidence that the proposed approach is more common.

Reply:
Dropping in one of many examples: https://github.com/golang/go/blob/go1.13.10/src/net/http/server.go#L822-L836